src/osd/ReplicatedPG.cc

```diff
@@ -5822,22 +5822,29 @@ int ReplicatedPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	  tracepoint(osd, do_osd_op_pre_omapgetkeys, soid.oid.name.c_str(), soid.snap.val, "???", 0);
 	  goto fail;
 	}
+	if (max_return > g_conf->osd_max_omap_entries_per_request) {
+	  max_return = g_conf->osd_max_omap_entries_per_request;
+	}
 	tracepoint(osd, do_osd_op_pre_omapgetkeys, soid.oid.name.c_str(), soid.snap.val, start_after.c_str(), max_return);
-	set<string> out_set;
 
+	bufferlist bl;
+	uint32_t num = 0;
 	if (oi.is_omap()) {
 	  ObjectMap::ObjectMapIterator iter = osd->store->get_omap_iterator(
 	    coll, ghobject_t(soid)
 	    );
 	  assert(iter);
 	  iter->upper_bound(start_after);
-	  for (uint64_t i = 0;
-	       i < max_return && iter->valid();
-	       ++i, iter->next(false)) {
-	    out_set.insert(iter->key());
+	  for (num = 0;
+	       num < max_return &&
+		 bl.length() < g_conf->osd_max_omap_bytes_per_request &&
+		 iter->valid();
+	       ++num, iter->next(false)) {
+	    ::encode(iter->key(), bl);
 	  }
 	} // else return empty out_set
-	::encode(out_set, osd_op.outdata);
+	::encode(num, osd_op.outdata);
```

Comment:
set<string> encodes as a u32 followed by N strings.  We encode the strings in bl above, encode teh count here (once we know it), then append bl.